commons/src/main/java/com/jagrosh/jdautilities/commons/utils/InverseAction.java

```diff
@@ -0,0 +1,616 @@
+    {
+        Member member = event.getMember();
+        boolean action = event.isGuildMuted();
+
+        return member.mute(!action);
+    }
+
+    /**
+     * @return An attempt to remove said override
+     */
+    public static AuditableRestAction<Void> of(PermissionOverrideCreateEvent event)
+    {
+        return event.getPermissionOverride().delete();
+    }
+
+    /**
+     * @return An attempt to add the override back
+     * @throws com.jagrosh.jdautilities.commons.utils.InverseAction.InversionException If the override's permission holder isn't cached
+     */
+    public static PermissionOverrideAction of(PermissionOverrideDeleteEvent event)
+    {
+        PermissionOverride deleted = event.getPermissionOverride();
+        IPermissionHolder holder = event.getPermissionHolder();
+        EnumSet<Permission> allowed = deleted.getAllowed();
+        EnumSet<Permission> denied = deleted.getDenied();
+
+        if (holder == null)
+            throw new InversionException("No permission holder in cache to use to copy deleted override");
+
+        return event.getChannel()
+                    .createPermissionOverride(holder)
+                    .setPermissions(allowed, denied);
+    }
+
+    /**
+     * @return An attempt to set rules to what they just were
+     */
+    public static PermissionOverrideAction of(PermissionOverrideUpdateEvent event)
+    {
+        PermissionOverride updated = event.getPermissionOverride();
+        EnumSet<Permission> allow = event.getOldAllow();
+        EnumSet<Permission> deny = event.getOldDeny();
+
+        return updated.getManager()
+                      .setAllow(allow).setDeny(deny);
+    }
+
+    /**
+     * @return An attempt to remove said emote
+     */
+    public static AuditableRestAction<Void> of(EmoteAddedEvent event)
+    {
+        return event.getEmote().delete();
+    }
+
+    /**
+     * @return An attempt to change the name back to what it just was
+     */
+    public static EmoteManager of(EmoteUpdateNameEvent event)
+    {
+        Emote emote = event.getEmote();
+        EmoteManager manager = emote.getManager();
+        String oldName = event.getOldName();
+
+        return manager.setName(oldName);
+    }
+
+    /**
+     * @return An attempt to change the roles to what they just were
+     */
+    public static EmoteManager of(EmoteUpdateRolesEvent event)
+    {
+        Emote emote = event.getEmote();
+        EmoteManager manager = emote.getManager();
+        HashSet<Role> oldRoles = new HashSet<>(event.getOldRoles());
+
+        return manager.setRoles(oldRoles);
+    }
+
+    /**
+     * @return An attempt to remove said role
+     */
+    public static AuditableRestAction<Void> of(RoleCreateEvent event)
+    {
+        return event.getRole().delete();
+    }
+
+    /**
+     * @return An attempt to add the role back
+     */
+    public static RoleAction of(RoleDeleteEvent event)
+    {
+        return event.getRole().createCopy();
+    }
```

Comment:
It should be made more clear which events are only "partially" reversible or if those should even be included. That is, you can create a copy of that role, but everything that still uses the old id is still broken and the members that used to have it don't have it anymore.
Same for e.g. voice channel delete.

I see it being mentioned in the summary javadoc, but it's still somewhat confusing.

Reply:
I wouldn't include inverse actions for deleting/removing events since you can't invert it fully.

Reply:
I believe it could be part of the documentation. I agree, my contract promises full reversion, but since that's not a promise I can keep while still wanting to implement the inverses to deletions, I can simply include it in the contract that must inversions regarding the removal of an entity will only be partial, and all immutable state from the original entity cannot possibly be copied